### PR TITLE
prevent serialization of symlinks by default

### PR DESCRIPTION
**Pull request by spencerschrock**
#### Summary

Adds a serializer initializer argument (`allow_symlink`) to control whether symlinks raise a `ValueError` during serialization. 

For rationale, see https://github.com/sigstore/model-transparency/issues/196#issuecomment-2245977345

#### Release Note

NONE

#### Documentation

NONE

### CODE REVIEW

File: model_signing/serialization/conftest.py [hunk 1]
```diff
@@ -102,3 +104,12 @@ def deep_model_folder(tmp_path_factory):
         file.write_text(f"This is file f{i}.")
 
     return model_root
+
+@pytest.fixture
+def symlink_model_file(tmp_path_factory: pytest.TempPathFactory) -> pathlib.Path:
+    """A model with a symlink to a single file."""
+    source_file = tmp_path_factory.mktemp("source") / "file"
+    source_file.write_bytes(test_support.KNOWN_MODEL_TEXT)
```
Comment (spencerschrock) [inline model_signing/serialization/conftest.py:112]: I could also use the sample_model_file fixture directly to create the source file

File: model_signing/serialization/serialize_by_file.py [hunk 2]
```diff
@@ -38,10 +41,18 @@ def check_file_or_directory(path: pathlib.Path) -> None:
...
 
     Raises:
-        ValueError: The path is neither a file or a directory.
+        ValueError: The path is neither a file or a directory, or the path
+          is a symlink and allow_symlinks is false.
     """
+    if not allow_symlinks and path.is_symlink():
+        raise ValueError(
+            f"Cannot use '{path}' because it is a symlink. This"
+            " behavior can be changed with allow_symlinks."
```
Comment (mihaimaruseac) [inline model_signing/serialization/serialize_by_file.py:54]:
```suggestion
            " behavior can be changed with `allow_symlinks`."
```

File: model_signing/serialization/serialize_by_file.py [hunk 3]
```diff
@@ -95,15 +107,18 @@ def __init__(
               hash individual files.
             max_workers: Maximum number of workers to use in parallel. Default
               is to defer to the `concurrent.futures` library.
+            allow_symlinks: Controls whether symlinks are serialized or raise
+              a ValueError. Default is to disallow symlinks.
```
Comment (mihaimaruseac) [inline model_signing/serialization/serialize_by_file.py:111]: Let's include a `Raises` section.

Also, nit: "Controls whether symbolic link are included. If a symlink is present but the flag is False (default) the serialization would raise an error."

File: model_signing/serialization/serialize_by_file.py [hunk 4]
```diff
@@ -284,13 +302,15 @@ def __init__(
             merge_hasher_factory: A callable that returns a
               `hashing.StreamingHashEngine` instance used to merge individual
               file digests to compute an aggregate digest.
+            allow_symlinks: Controls whether symlinks are serialized or raise
```
Comment (mihaimaruseac) [inline model_signing/serialization/serialize_by_file.py:305]: Similar here

File: model_signing/serialization/serialize_by_file_shard.py [hunk 5]
```diff
@@ -104,9 +105,12 @@ def __init__(
               the shard.
             max_workers: Maximum number of workers to use in parallel. Default
               is to defer to the `concurrent.futures` library.
+            allow_symlinks: Controls whether symlinks are serialized or raise
+              a ValueError. Default is to disallow symlinks.
```
Comment (mihaimaruseac) [inline model_signing/serialization/serialize_by_file_shard.py:109]: And here

File: model_signing/serialization/serialize_by_file_shard.py [hunk 6]
```diff
@@ -117,7 +121,9 @@ def __init__(
     def serialize(self, model_path: pathlib.Path) -> manifest.Manifest:
         # TODO: github.com/sigstore/model-transparency/issues/196 - Add checks
         # to exclude symlinks if desired.
-        serialize_by_file.check_file_or_directory(model_path)
+        serialize_by_file.check_file_or_directory(
```
Comment (mihaimaruseac) [inline model_signing/serialization/serialize_by_file_shard.py:124]: I think we can exclude the TODO above now.

File: model_signing/serialization/serialize_by_file_shard.py [hunk 7]
```diff
@@ -243,8 +252,10 @@ def __init__(
               individual file shard digests to compute an aggregate digest.
             max_workers: Maximum number of workers to use in parallel. Default
               is to defer to the `concurent.futures` library.
+            allow_symlinks: Controls whether symlinks are serialized or raise
+              a ValueError. Default is to disallow symlinks.
```
Comment (mihaimaruseac) [inline model_signing/serialization/serialize_by_file_shard.py:256]: Here too for including a `Raises` section

File: model_signing/serialization/serialize_by_file_shard_test.py [hunk 8]
```diff
@@ -85,7 +85,9 @@ def test_known_models_small_shards(self, request, model_fixture_name):
 
         # Compute model manifest (act)
         serializer = serialize_by_file_shard.DigestSerializer(
-            self._hasher_factory_small_shards, memory.SHA256()
+            self._hasher_factory_small_shards,
+            memory.SHA256(),
+            allow_symlinks=True
```
Comment (mihaimaruseac) [inline model_signing/serialization/serialize_by_file_shard_test.py:90]:
```suggestion
            allow_symlinks=True,
```

**Review by mihaimaruseac — COMMENTED** [review]
Thank you! This LGTM.

Added a request to change the test fixture to be a directory and then some minor nits here and there.

File: model_signing/conftest.py [hunk 9]
```diff
@@ -102,3 +104,12 @@ def deep_model_folder(tmp_path_factory):
         file.write_text(f"This is file f{i}.")
 
     return model_root
+
+@pytest.fixture
+def symlink_model_file(tmp_path_factory: pytest.TempPathFactory) -> pathlib.Path:
```
[the remark this answers is not in the document]
Comment (spencerschrock) [reply model_signing/conftest.py]: 475edd52d4595a0f6b77d0fc119f2e79c8289c87

File: model_signing/serialization/serialize_by_file.py (hunk 3 above)
Comment (spencerschrock) [reply model_signing/serialization/serialize_by_file.py]: Addressed the nit.
With regards to the `Raises` section, the __init__() doesn't raise the error, is it still relevant to include?

Comment (mihaimaruseac) [reply model_signing/serialization/serialize_by_file.py]: Oh, I missed this was on the `__init__`. It shouldn't be there, only when this gets raised.

**Review by mihaimaruseac — DISMISSED** [review]
Thank you!

File: model_signing/serialization/serialize_by_file.py (hunk 3 above)
Comment (spencerschrock) [reply model_signing/serialization/serialize_by_file.py]: I'll add this to the various serializer methods

Comment (mihaimaruseac) [reply model_signing/serialization/serialize_by_file.py]: Sorry, I made it worse when I tried to resolve the conflict from the web interface. Apparently that doesn't insert a newline at the end of the file 🤦

Comment (spencerschrock) [reply model_signing/serialization/serialize_by_file.py]: Force pushed away :)

Please take another look at the docstrings in 053de6f43037af4224f8453f295a030ee3352b8d

**Review by mihaimaruseac — APPROVED** [review]
Nice! Thank you!

File: model_signing/serialization/serialize_by_file.py (hunk 3 above)
Comment (mihaimaruseac) [reply model_signing/serialization/serialize_by_file.py]: Nice! You covered the ones that are a method call away (via `super()`, etc.). Well done!

**Pull request merged**